commons/com.b2international.commons/src/com/b2international/commons/io/PathUtils.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright 2023 B2i Healthcare, http://b2ihealthcare.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.b2international.commons.io;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Sets.newHashSet;
+import static java.util.stream.Collectors.toList;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.*;
+import java.util.*;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+import com.google.common.base.Strings;
+
+/**
+ * @since 7.22.0
+ */
+public class PathUtils {
+
+	/**
+	 * Delete the specified directory and all files recursively
+	 */
+	public static void deleteDirectory(final Path directory) throws IOException {
+		deleteDirectory(directory, Set.of());
+	}
+
+	/**
+	 * Clean the specified directory but keep the directory itself
+	 */
+	public static void cleanDirectory(final Path directory) throws IOException {
+		deleteDirectory(directory, Set.of(directory.toString()));
+	}
+
+	/**
+	 * Delete all files within the specified directory except the exclusions specified. If there are none, the directory is deleted as well.
+	 * The exclusions must be declared as full paths.
+	 */
+	public static void deleteDirectory(final Path directory, Collection<String> exclusions) throws IOException {
+
+		checkArgument(Files.isDirectory(directory), "The specified path is not a directory");
+
+		final List<Path> filteredPaths = Files.walk(directory)
+				.sorted(Comparator.reverseOrder())
+				.filter(path -> exclusions.stream().noneMatch(p -> p.equals(path.toString())))
```

Comment:
If `exclusions` held a `Collection` of `Path` instances, we the comparisons would honor file system specific conditions, eg. case sensitivity, according to `Path#equals`'s javadoc. Which may or may not be a good thing. 😄

Reply:
That's a good idea, will change it accordingly!